diff --git a/libs/milvus/langchain_milvus/vectorstores/milvus.py b/libs/milvus/langchain_milvus/vectorstores/milvus.py
index a73f81c..235a047 100644
--- a/libs/milvus/langchain_milvus/vectorstores/milvus.py
+++ b/libs/milvus/langchain_milvus/vectorstores/milvus.py
@@ -31,6 +31,7 @@ from pymilvus import (
     MilvusException,
     RRFRanker,
     WeightedRanker,
+    Function
 )
 from pymilvus.orm.types import infer_dtype_bydata  # type: ignore
 
@@ -141,6 +142,34 @@ def maximal_marginal_relevance(
     return idxs
 
 
+def check_model_ranker_params(ranker_params: dict):
+    provider = ranker_params.get("provider", "")
+    if provider not in ["vllm", "tei", "cohere", "voyageai", "siliconflow"]:
+        raise ValueError(
+            f"Provider {provider} is not supported. "
+            "Please use one of the following providers: vllm, tei, cohere, voyageai, siliconflow."
+        )
+    if "queries" not in ranker_params:
+        raise ValueError(
+            f"Provider {provider} requires a queries parameter. "
+            "Please provide a queries parameter in the ranker_params."
+        )
+    if provider in ["vllm", "tei"]:
+        if "endpoint" not in ranker_params:
+            raise ValueError(
+                f"Provider {provider} requires an endpoint. "
+                "Please provide an endpoint in the ranker_params."
+            )
+    if provider in ["cohere", "voyageai", "siliconflow"]:
+        if "model_name" not in ranker_params:
+            raise ValueError(
+                f"Provider {provider} requires a model_name parameter. "
+                "Please provide a model_name parameter in the ranker_params."
+            )
+    if "reranker" not in ranker_params or ranker_params["reranker"] != "model":
+        ranker_params["reranker"] = "model"
+
+
 EmbeddingType = Union[Embeddings, BaseSparseEmbedding]
 T = TypeVar("T")
 
@@ -1601,7 +1630,7 @@ class Milvus(VectorStore):
         param: Optional[dict | list[dict]] = None,
         expr: Optional[str] = None,
         fetch_k: Optional[int] = 4,
-        ranker_type: Optional[Literal["rrf", "weighted"]] = None,
+        ranker_type: Optional[Literal["rrf", "weighted", "model"]] = None,
         ranker_params: Optional[dict] = None,
         timeout: Optional[float] = None,
         **kwargs: Any,
@@ -2367,9 +2396,9 @@ class Milvus(VectorStore):
 
     def _create_ranker(
         self,
-        ranker_type: Optional[Literal["rrf", "weighted"]],
+        ranker_type: Optional[Literal["rrf", "weighted", "model"]],
         ranker_params: dict,
-    ) -> Union[WeightedRanker, RRFRanker]:
+    ) -> Union[WeightedRanker, RRFRanker, Function]:
         """A Ranker factory method"""
         default_weights = [1.0] * len(self._as_list(self._vector_field))
         if not ranker_type:
@@ -2383,6 +2412,14 @@ class Milvus(VectorStore):
             if k:
                 return RRFRanker(k)
             return RRFRanker()
+        elif ranker_type == "model":
+            check_model_ranker_params(ranker_params)
+            return Function(
+                name=f"{ranker_params['provider']}_semantic_ranker",
+                input_field_names=["text"],
+                function_type=FunctionType.RERANK,
+                params=ranker_params,
+            )
         else:
             logger.error(
                 "Ranker %s does not exist. "
@@ -2390,6 +2427,7 @@ class Milvus(VectorStore):
                 ranker_type,
                 "weighted",
                 "rrf",
+                "model",
             )
             raise ValueError("Unrecognized ranker of type %s", ranker_type)
 
@@ -2702,7 +2740,7 @@ class Milvus(VectorStore):
         param: Optional[dict | list[dict]] = None,
         expr: Optional[str] = None,
         fetch_k: Optional[int] = 4,
-        ranker_type: Optional[Literal["rrf", "weighted"]] = None,
+        ranker_type: Optional[Literal["rrf", "weighted", "model"]] = None,
         ranker_params: Optional[dict] = None,
         timeout: Optional[float] = None,
         **kwargs: Any,
